proofs/salts.go

```diff
@@ -0,0 +1,107 @@
+package proofs
+
+import (
+	"fmt"
+	"github.com/golang/protobuf/proto"
+	"reflect"
+	"strings"
+)
+
+// FillSalts is a helper message that iterates over all fields in a proto.Message struct and fills them with 32 byte
+// random values.
+//
+// This method will fail if there are any fields of type other than []byte (bytes in protobuf) in the
+// message leaf.
+//
+// The `saltsMessage` protobuf value provided will need to add one extra field for each field that is defined as `repeated`.
+//
+// The suffix of the slice field needs to end in: `Length` or customized as described above.
+func FillSalts(dataMessage, saltsMessage proto.Message) (err error) {
+	dataMessageValue := reflect.Indirect(reflect.ValueOf(dataMessage))
+	saltsMessageValue := reflect.Indirect(reflect.ValueOf(saltsMessage))
+
+	for i := 0; i < saltsMessageValue.NumField(); i++ {
+		saltsField := saltsMessageValue.Field(i)
+		saltsType := saltsField.Type()
+		valueField := dataMessageValue.FieldByName(saltsMessageValue.Type().Field(i).Name)
+
+		if saltsType.Kind() == reflect.Ptr {
+			saltsType = saltsType.Elem()
+		}
+
+		if strings.HasPrefix(saltsMessageValue.Type().Field(i).Name, "XXX_") {
+			continue
+		}
+
+		if saltsType == reflect.TypeOf([]uint8{}) {
+			err = handleFillSaltsValue(saltsField)
+		} else if saltsType.Kind() == reflect.Slice {
+			err = handleFillSaltsSlice(saltsType, saltsField, valueField)
+		} else if saltsType.Kind() == reflect.Struct {
+			err = handleFillSaltsStruct(saltsField, saltsType, valueField)
+		} else {
+			return fmt.Errorf("Invalid type (%s) for field", reflect.TypeOf(saltsField.Interface()).String())
+		}
+
+		if err != nil {
+			return
```

Comment:
Why are we not returning the error here?

Reply:
@mikiquantum do you know why? I just moved this code from A to B.

Reply:
Error is implicitly returned by named returns.